wnbd-client/client.cpp

```diff
@@ -0,0 +1,333 @@
+#include "client.h"
+#include "cmd.h"
+#include "usage.h"
+#include "wnbd.h"
+#include "version.h"
+
+#include <boost/exception/diagnostic_information.hpp>
+
+#include <iostream>
+#include <iomanip>
+
+namespace po = boost::program_options;
+using namespace std;
+
+vector<Client::Command*> Client::commands;
+
+Client::Command* Client::get_command(string name) {
+    for (Client::Command *command : Client::commands) {
+        if (command->name == name)
+            return command;
+        for (auto alias: command->aliases) {
+            if (alias == name)
+                return command;
+        }
+    }
+
+    return nullptr;
+}
+
+// Make sure to pass the EXACT SAME type that was used when defining the option,
+// otherwise Boost will throw an exception.
+template <class T>
+T safe_get_param(const po::variables_map& vm, string name, T default_val = T()) {
+    if (vm.count(name)) {
+        return vm[name].as<T>();
+    }
+    return default_val;
+}
+
+void Client::get_common_options(po::options_description &options) {
+    options.add_options()
+        ("debug", po::bool_switch(), "Enable debug logging.");
+}
+
+void handle_common_options(po::variables_map &vm) {
+    bool debug = safe_get_param<bool>(vm, "debug");
+
+    WnbdLogLevel log_level = debug ? WnbdLogLevelDebug : WnbdLogLevelInfo;
+    WnbdSetLogLevel(log_level);
+}
+
+DWORD Client::execute(int argc, const char** argv) {
+    vector<string> args;
+    args.insert(args.end(), argv + 1, argv + argc);
+
+    // The first must argument must be the command name.
+    if (args.size() == 0) {
+        args.push_back("help");
+    }
+    string command_name = args[0];
+
+    Client::Command* command = get_command(command_name);
+    if (!command) {
+        cerr << "Unknown command: " << command_name << endl << endl
+             << "See wnbd-client --help." << endl;
+        return ERROR_INVALID_PARAMETER;
+    }
+    // Remove the command from the list of arguments.
+    args.erase(args.begin());
+
+    try {
+        po::positional_options_description positional_opts;
+        po::options_description named_opts;
+
+        if (command->get_options) {
+            command->get_options(positional_opts, named_opts);
+        }
+        get_common_options(named_opts);
+
+        po::variables_map vm;
+        po::store(po::command_line_parser(args)
+            .options(named_opts)
+            .positional(positional_opts)
+            .run(), vm);
+        po::notify(vm);
+
+        handle_common_options(vm);
+        return command->execute(vm);
+    }
+    catch (po::required_option& e) {
+        cerr << "wnbd-client: " << e.what() << endl;
```

Comment:
Does it make sense to save the binary name inside the `main` function and reuse it when needed?

Reply:
Another idea would be to define a macro, something like CLIENT_NAME but I'd rather keep things simple for now.